src/api/c/pinverse.cpp

```diff
@@ -0,0 +1,170 @@
+    // very small non-zero values though (< relTol), in order to avoid very
+    // large reciprocals
+    double relTol = tol * static_cast<double>(max(M, N))
+                        * reduce_all<af_max_t, Tr, Tr>(sVec);
+    Array<T> relTolArr = createValueArray<T>(sVecCast.dims(), scalar<T>(relTol));
+    Array<T> ones = createValueArray<T>(sVecCast.dims(), scalar<T>(1.));
+    Array<T> sVecRecip = arithOp<T, af_div_t>(ones, sVecCast, sVecCast.dims());
+    Array<char> cond = logicOp<T, af_ge_t>(sVecCast, relTolArr,
+                                           sVecCast.dims());
+    Array<T> zeros = createValueArray<T>(sVecCast.dims(), scalar<T>(0.));
+    sVecRecip = createSelectNode<T>(cond, sVecRecip, zeros, sVecRecip.dims());
+    sVecRecip.eval();
+
+    // Make s vector into s pinverse array
+    Array<T> sPinv = diagCreate<T>(sVecRecip, 0);
+
+    Array<T> uT = transpose(u, true);
+
+    // Crop v and u* for final matmul later based on s+'s size, because
+    // sVec produced by svd() has minimal dim length (no extra zeroes).
+    // Thus s+ produced by diagCreate() will have minimal dims as well,
+    // and v could have an extra dim0 or u* could have an extra dim1
+    if (v.dims()[1] > sPinv.dims()[0]) {
+        vector<af_seq> seqs = {
+            {0., static_cast<double>(v.dims()[0] - 1), 1.},
+            {0., static_cast<double>(sPinv.dims()[0] - 1), 1.}
+        };
+        v = createSubArray<T>(v, seqs);
+    }
+    if (uT.dims()[0] > sPinv.dims()[1]) {
+        vector<af_seq> seqs = {
+            {0., static_cast<double>(sPinv.dims()[1] - 1), 1.},
+            {0., static_cast<double>(uT.dims()[1] - 1), 1.}
+        };
+        uT = createSubArray<T>(uT, seqs);
+    }
+
+    Array<T> out = matmul<T>(matmul<T>(v, sPinv, AF_MAT_NONE, AF_MAT_NONE),
+                             uT, AF_MAT_NONE, AF_MAT_NONE);
+
+    return out;
+}
+
+template<typename T>
+static inline af_array pinverse(const af_array in, const double tol)
+{
+    Array<T> inArray = getArray<T>(in);
+    uint batchSize = inArray.dims()[2];
+
+    if (inArray.ndims() < 3) {
+        return getHandle(pinverseSvd<T>(getArray<T>(in), tol));
+    }
+    else {
+        vector<Array<T> > outputs;
+        for (int i = 0; i < batchSize; ++i) {
+            vector<af_seq> seqs = {
+                {0., static_cast<double>(inArray.dims()[0] - 1), 1.},
+                {0., static_cast<double>(inArray.dims()[1] - 1), 1.},
+                {static_cast<double>(i), static_cast<double>(i), 1.}
+            };
+            Array<T> inSlice = createSubArray<T>(inArray, seqs);
+            outputs.push_back(pinverseSvd<T>(inSlice, tol));
+        }
+        Array<T> mergedOuts = join<T>(2, outputs);
+        return getHandle(mergedOuts);
+    }
+}
+
+af_err af_pinverse(af_array *out, const af_array in, const double tol,
+                   const af_mat_prop options)
+{
+    try {
+        const ArrayInfo& i_info = getInfo(in);
+
+        if (i_info.ndims() > 3) {
```

Comment:
Is this check still necessary? It looks like you are able to handle batched operations.

Reply:
Even with batching, I don't handle arrays with 4 dimensions, just 3. Not sure how to go about that case, maybe we could talk more about this.

Reply:
You can flatten the 3rd and 4th dimension and then perform the inverse. so reshape it to (M, N, d3 * d4, 1)

Reply:
Welp I guess I could extend batching to 4 dimensions - I simply have to enclose my current batching logic in another for loop for each 3D slice, collect all of them, and finally join them along the 4th dimension